[ADD] add a new module to change the original back-order logic of Odoo

duanyp1991 opened the pull request:
in chained locations introducing true back-to-back orders.
 Changes to be committed:

```
new file:   stock_back2back_order_proc/__init__.py
new file:   stock_back2back_order_proc/__openerp__.py
new file:   stock_back2back_order_proc/stock.py
new file:   stock_back2back_order_proc/stock_view.xml
new file:   stock_back2back_order_proc/test/test_back2back_order_proc.yml
```

Commit:
[ADD] add a new module to change the original back-order logic of Odoo
in chained locations introducing true back-to-back orders.
 Changes to be committed:
	new file:   stock_back2back_order_proc/__init__.py
	new file:   stock_back2back_order_proc/__openerp__.py
	new file:   stock_back2back_order_proc/stock.py
	new file:   stock_back2back_order_proc/stock_view.xml
	new file:   stock_back2back_order_proc/test/test_back2back_order_proc.yml

Commit:
[FIX] update copyright information.
 Changes to be committed:
	modified:   stock_back2back_order_proc/__init__.py
	modified:   stock_back2back_order_proc/stock.py

gurneyalex commented on stock_back2back_order_proc/__openerp__.py:
```diff
@@ -0,0 +1,126 @@
+
+STANDARD ODOO BACK-TO-BACK ORDER BEHAVIOR:
+Original behavior is not fully suitable to handle back-to-back backorders
+    process (check "back-to-back orders comparison.pdf"):
```
where is this file?

duanyp1991 replied:
Thanks for your advise,Alexandre! Sorry for this late reply since it was during Chinese New Year holiday. I'll start working on this today.:)

gurneyalex commented on stock_back2back_order_proc/__openerp__.py:
```diff
@@ -0,0 +1,126 @@
+STANDARD ODOO BACK-TO-BACK ORDER BEHAVIOR:
+Original behavior is not fully suitable to handle back-to-back backorders
+    process (check "back-to-back orders comparison.pdf"):
+eg: Let's take the following example to understand the implemented difference:
```
s/eg://

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+            help=_("In 'Retention mode' the system wait for the\
+                  whole quantity before the stuff is processed.\n" +
+                   "In 'Thru mode' the shipped quantity is processed regardless\
+                  of the ordered quantity.")
```
stylistic note: you don't need the backquotes or the `+` here, you can rely on the automatic joint of string litterals in Python. This is cleaner because it does not include the indentation spaces in the string

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+            [('retention', _('Retention Mode')), ('thru', _('Thru mode'))],
+            _('Retention Mode'),
+            required=True,
+            help=_("In 'Retention mode' the system wait for the\
```
s/wait/waits/

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+            _('Retention Mode'),
+            required=True,
+            help=_("In 'Retention mode' the system wait for the\
+                  whole quantity before the stuff is processed.\n" +
```
s/stuff is/products/

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+        '''
+        if not move_obj:
+            move_obj = self.pool.get('stock.move')
+        move_tbc = move_obj.browse(cr, uid, move_id, context, move_obj)
```
what does 'tbc' mean in this context?

gurneyalex replied:
You are passing `move_obj` to the `list_class` argument of `browse()` and I am pretty sure this is _not_ what you want to do. I am also surprised that this did not crash when you tested your code, because it really looks like a :bug:

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+        # If there is move_dest_id in the chain
+        if move_tbc.move_dest_id:
+            move_chain = self.get_move_chain(
+                cr, uid, move_tbc.move_dest_id.id, context)
```
you are not passing `move_ob`j in your recursive call. I think this is fine and you shoud remove the argument as well as the test on line 56.

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+    _inherit = "stock.picking"
+
+    def get_move_chain(self, cr, uid, move_id, context=None, move_obj=False):
+        '''Recursively get the chained moves
```
Generally speaking, I'd rather have an iterative implementation of this than a recursive implementation. This is much more efficient in Python (while loop vs. function call, you can share local variables between calls, no risk of `MaxRecursionDepthExceeded` exceptions, and Python has no tail optimisation for recursion)

In this specific case, recursion might be ok, because the chains are generally short, but this is something you may want to keep in mind.

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+            move_chain = []
+
+        move_chain.append(move_tbc)
+        return move_chain
```
you are returning a list of browse records, not a list of ids, so this method must not be exposed to xmlrpc. You should rename it `_get_move_chain`

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+        @return list of the chained moves
+        '''
+        if not move_obj:
+            move_obj = self.pool.get('stock.move')
```
use `self.pool['stock.move']`

duanyp1991 replied:
@gurneyalex , For performance reason? Do you think we should use this for all cases?

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+                # TODO handle this kind of exception.
+                raise orm.except_orm(
+                    _('Data Error'),
+                    _('There is no picking relates to this move: ') +
```
s/relates/related/

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+        for move in all_chained_moves:
+            if not move.picking_id:
+                # TODO handle this kind of exception.
+                raise orm.except_orm(
```
do we really want to completely stop all processing in this case?

duanyp1991 replied:
You are right, we might not need to block whole process.

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+                raise orm.except_orm(
+                    _('Data Error'),
+                    _('There is no picking relates to this move: ') +
+                    '%s' % move.name)
```
you should keep the `%s` inside the translated string, as it allow translators to put the translation elsewhere than in the end of the string.

LeartS replied:
relate**s** -> relate**d**

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+                    _('Data Error'),
+                    _('There is no picking relates to this move: ') +
+                    '%s' % move.name)
+            if move.picking_id.id and (move.picking_id.id not in new_picks):
```
you can get rid of the first part of the test: you've just raised an exception in that case.

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+                    '%s' % move.name)
+            if move.picking_id.id and (move.picking_id.id not in new_picks):
+                pick_tbc = self.browse(cr, uid, move.picking_id.id, context)
+                new_note = ((pick_tbc.note if pick_tbc.note else '') +
```
use `pick_tbc.note or u''` here

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+            if move.picking_id.id and (move.picking_id.id not in new_picks):
+                pick_tbc = self.browse(cr, uid, move.picking_id.id, context)
+                new_note = ((pick_tbc.note if pick_tbc.note else '') +
+                            ' Copy of stock.pick[%d].') % move.picking_id.id
```
this string needs to be translated. 

and you probably need to add a `\n` somewhere.

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+
+    def copy_move_chain(
+            self, cr, uid, move_id, product_qty, new_picks, context=None, move_obj=False):
+        '''Recursively copy the chained move until a location in retention mode or the end.
```
same remark about recursion.

gurneyalex replied:
s/move/moves/

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+
+        my_picking_id = new_picks.get(move_tbc.picking_id.id, False)
+
+        new_note = ((move_tbc.note if move_tbc.note else '') + ' Copy of stock.move[%d].') % move_id
```
same remarks as on lines 99-100

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+        return new_move_id
+
+    def update_move_chain_pick(self, cr, uid, move_id, vals, new_picks, context=None):
+        '''Recursively update the new chained move with the new related picking
```
recursively you say? I don't think so :smile_cat:

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+        return True
+
+    def update_move_chain(self, cr, uid, move_id, vals, context=None):
+        '''Recursively update the old chained move by the first move id
```
nope, this is not a recursive function.

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+            context = dict(context)
+        res = {}
+        move_obj = self.pool.get('stock.move')
+        uom_obj = self.pool.get('product.uom')
```
`self.pool[model.name]`

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+
+    # FIXME: needs refactoring, this code is partially duplicated in stock_move.
+    # do_partial()!
+    def do_partial(self, cr, uid, ids, partial_datas, context=None):
```
needs fixing: this code does not call the super() implementation of do_partial
- this needs to be noted in the docstrving (and in the module documentation)
- I assume most of the code below is copied from the original implementaiton. Please highlight the changed parts with notes in comments.

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+            move_dest_id = self.copy_move_chain(
+                cr, uid, move_tbc.move_dest_id.id, product_qty, context)
+
+        new_note = ((move_tbc.note if move_tbc.note else '') + ' Copy of stock.move[%d].') % move_id
```
same remarks as on line 99-100

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+    _inherit = "stock.move"
+
+    def copy_move_chain(self, cr, uid, move_id, product_qty, context=None):
+        '''Recursively copy the chained move until a location in retention mode or the end.
```
recursion again...

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+        return new_move_id
+
+    def update_move_chain(self, cr, uid, move_id, vals, context=None):
+        '''Recursively update the chained move by the first move id
```
no recursion in this method AFAICT

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+        '''
+
+        if isinstance(move_id, list):
+            move_id = move_id[0]
```
please `assert len(move_id) == 1`

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+
+        if isinstance(move_id, list):
+            move_id = move_id[0]
+        ids = [move_id]
```
remove this

gurneyalex commented on stock_back2back_order_proc/stock.py:
```diff
@@ -0,0 +1,552 @@
+        return True
+
+    def do_partial(self, cr, uid, ids, partial_datas, context=None):
+        """ Makes partial pickings and moves done.
```
same remark as for the previous `do_partial` reimplementation

gurneyalex commented on stock_back2back_order_proc/stock_view.xml:
```diff
@@ -0,0 +1,17 @@
+    <data>
+		<record model="ir.ui.view" id="view_location_form_back2back">
+		    <field name="name">stock.location.form.back2back</field>
+            <field name="model">stock.location</field>
```
please fix intentation

gurneyalex commented:
Other points to fix
- [ ] travis is red
- [ ] missing `i18n/stock_back2back_order_proc.pot`

yvaucher commented:
@duanyp1991 Don't forget to settup your email on your local git config so Github will recognize your commits

Commit:
[IMP]
    * iterative implementation instead of recursive methods
    * add new file: Back-to-back orders comparison.pdf
    * add new file: en_US.po
    * change a few typos
    * make some of the warning msg translatable
    * improve yml tests.
    * for method do_partial:
        - add comments to mark changes
        - add warning in the __openerp__.py
    * using automatic joint of string litterals instead of '+' operation
    * changing from self.pool.get(model_name) to self.pool[model_name]
 Changes to be committed:
	modified:   stock_back2back_order_proc/__openerp__.py
	new file:   stock_back2back_order_proc/doc/back-to-back orders comparison.pdf
	new file:   stock_back2back_order_proc/i18n/en_US.po
	modified:   stock_back2back_order_proc/stock.py
	modified:   stock_back2back_order_proc/stock_view.xml
	modified:   stock_back2back_order_proc/test/test_back2back_order_proc.yml

gurneyalex commented:
@duanyp1991 can you rename the file en_US.po to stock_back2back_order_proc.pot?

gurneyalex commented:
@duanyp1991 do you have still interest in this PR?

pedrobaeza commented:
As there's no answer, I close. Feel free to reopen in case there's interest.
